### DISCUSSION TITLE
Webhooks and SSL verification

### DISCUSSION THREAD

**Discussion by kvanderkamp** [community/community#23434] [New to GitHub]
<p>We’re running Jenkins with HTTPS acces using a certificate from Let’s Encrypt.</p>
<p>The following procedure was used to generate and install the certificate:</p>
<p><a href="https://github.com/hughperkins/howto-jenkins-ssl/blob/master/letsencrypt.md" rel="nofollow noopener">https://github.com/hughperkins/howto-jenkins-ssl/blob/master/letsencrypt.md</a></p>
<p>Connecting from my browser works as it should, the browser indicates the connection is secured and verified by Let’s Encrypt.</p>
<p>After updating the webhook in our GitHub repository from connecting over HTTP to HTTPS, GitHub fails to deliver the webhook payload with the following message:</p>
<p>“We couldn’t deliver this payload: <strong>Peer certificate cannot be authenticated with given CA certificates</strong>”</p>
<p>How come the certificate works in my browser, but doesn’t work for webhooks?</p>

**Comment by lee-dohm** [comment]
<p>All HTTPS certificates are issued by some entity. These certificates are often signed by a Certificate Authority or CA. That signature’s authenticity is verified against the CA’s certificate which could also be signed by another CA. This can continue in a chain until you either reach the end of the chain or reach certificate signed by a CA that the system implicitly trusts (also known as a “root CA”).</p>
<p>For example, you can see the list of root CAs recognized by your system on macOS by using the Keychain Access app under the System Roots keychain:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/a/ab34fd72ad24c34efafb823d9d96710aafb256aa.png" data-download-href="/uploads/short-url/oqzb2tOL6LNMg5lLDCPjnblYkRQ.png?dl=1" title="Screen Shot 2019-01-25 at 10.45.18 AM.png"><img src="https://user-images.githubusercontent.com/1038121/181076905-8445fe22-ea41-448b-a8b4-8f43a0785f32.png" alt="Screen Shot 2019-01-25 at 10.45.18 AM.png" data-base62-sha1="oqzb2tOL6LNMg5lLDCPjnblYkRQ" width="675" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/a/ab34fd72ad24c34efafb823d9d96710aafb256aa_2_675x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/a/ab34fd72ad24c34efafb823d9d96710aafb256aa_2_1012x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/a/ab34fd72ad24c34efafb823d9d96710aafb256aa_2_1350x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/a/ab34fd72ad24c34efafb823d9d96710aafb256aa_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2019-01-25 at 10.45.18 AM.png</span><span class="informations">2534×1876 1.08 MB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Different systems can have different lists of trusted root CAs. This is why your browser and our webhook system can have different results when attempting to create an HTTPS connection to the same system.</p>
<p>On the other hand, we haven’t had any reports of Let’s Encrypt certificates not being honored by our webhook system. You may want to ensure that your system is serving the entire certificate chain when establishing HTTPS connections.</p>
<p>If you continue to run into problems, you may want to write into private support at <a href="https://github.com/contact">https://github.com/contact</a> so they can take a deeper look at your configuration.</p>
<p>I hope that helps!</p>

**Comment by cloutiertyler** [comment]
<p>Were you able to determine if this was this issue?</p>

**Comment by kvanderkamp** [comment]
<p>I still haven’t been able to determine what is the issue.</p>
<p><a href="https://whatsmychaincert.com/" rel="nofollow noopener">https://whatsmychaincert.com/</a> indicates my server is misconfigured and provides a chain it should be using. The suggested chain only contains <strong>Let’s Encrypt Authority X3</strong> , whereas the <code>fullchain.pem</code> served by Jenkins contains both my own domain cert and the Let’s Encrypt Authority X3 cert. When using the suggested chain neither my browser nor GitHub successfully connect over HTTPS.</p>
<p>If this persists I will write into private support.</p>

**Comment by cloutiertyler** [comment]
<p>Ah, interesting, I’m having the exact same issue! We’re using fullchain.pem and if we use the suggested chain neither the browser nor GitHub successfully connect over HTTPS. Not sure how to proceed at this point.</p>
<p>If you do make progress through private support, please update here if possible.</p>

**Comment by aasoni** [comment, marked as answer]
<p>We solved this issue by starting Jenkins using the two flags: --httpsKeyStore and --httpsKeyStorePassword instead of --httpsCertificate and --httpsPrivateKey.</p>
<p>In the Jenkins startup logs it mentions that httpsKeyStore is the preferred method. It seems that using the old flags Jenkins was not serving the correct certificate chain.</p>
<p>You can generate the keystore using:</p>
<p>openssl pkcs12 -inkey privkey.pem -in fullchain.pem -export -out keys.pkcs12<br>
keytool -importkeystore -srckeystore keys.pkcs12 -srcstoretype pkcs12 -destkeystore keystore</p>
<p>After starting jenkins using this generated keystore we no longer had issues with github webhooks (also <a href="https://whatsmychaincert.com" rel="nofollow noopener">https://whatsmychaincert.com</a> showed that the correct chain was being served)</p>

**Comment by kvanderkamp** [comment]
<p>This solution works great. Thank you!</p>
<p>Note that Jenkins won’t start when generating the keys.pkcs12 and keystore files using two different passwords. Using the same password for both files solves this problem.</p>

**Comment by kvanderkamp** [comment]
<p>Someone posted a solution to this problem, which worked for me. You may want to give it a try.</p>
<p>Thanks for your help :slight_smile:</p>

**Comment by sicute** [comment]
<p>Thanks to give solustion but how when using jenkins in kubernetes any clue running customer jenkins with option   --httpsKeyStore and --httpsKeyStorePassword</p>
